src/ObjCRuntime/Runtime.cs

```diff
@@ -944,8 +944,12 @@ static Delegate CreateBlockProxy (MethodInfo method, IntPtr block)
 #endif
 		static Delegate GetDelegateForBlock (IntPtr methodPtr, Type type)
 		{
-			if (block_to_delegate_cache == null)
-				block_to_delegate_cache = new Dictionary<IntPtrTypeValueTuple, Delegate> ();
+			if (block_to_delegate_cache == null) {
+				lock (lock_obj) {
+					if (block_to_delegate_cache == null)
+						block_to_delegate_cache = new Dictionary<IntPtrTypeValueTuple, Delegate> ();
+				}
+			}
 
 			// We do not care if there is a race condition and we initialize two caches
```

Comment:
This comment is not accurate anymore, and should be removed.

Reply:
Looks like the comment is about this line, so it is still accurate.
https://github.com/xamarin/xamarin-macios/blob/70ef53f5c5f26ee2d8cff84870f42ec9db6a69e6/src/ObjCRuntime/Runtime.cs#L964

Reply:
Interesting how we understood the comment differently :smile:

My take is that the code was referring to the fact that the `block_to_delegate_cache` creation was not synchronized ("we initialize two _caches_"), so that there could end up being multiple instances of the `block_to_delegate_cache` dictionary, and that all but one would eventually end up discarded (and then it turned out that this wasn't safe, as you found out).

However, the comment is still accurate about the fact that we may end up calling `Marshal.GetDelegateForFunctionPointer` multiple times for the same method pointer (which is fine).